torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -322,46 +362,78 @@ std::shared_ptr<::gloo::transport::Device> ProcessGroupGloo::
 std::shared_ptr<::gloo::transport::Device> ProcessGroupGloo::
     createDeviceForHostname(const std::string& hostname) {
   ::gloo::transport::tcp::attr attr;
+  attr.hostname = hostname;
+  TORCH_CHECK(
+      doesHostnameResolveToUsableAddress(attr.hostname),
+      "Cannot resolve ",
+      hostname,
+      " to a (local) address");
+  return ::gloo::transport::tcp::CreateDevice(attr);
```

Comment:
what does `CreateDevice` do in tcp and uv if the hostname is not valid?

Reply:
It crashes saying it cannot resolve the hostname or bind to the address. 

This is also what I want to prevent with this commit. Instead, we can either fall back to using the loopback interface when using the default constructor, or raise a runtime error with TORCH_CHECK prior to calling `CreateDevice` when using the non-default constructor.